### PR TITLE
[BBC] Make additional fields visible on the UI via configuration

### PR DESCRIPTION
**Pull request by ochiengolanga**
## What does this change?
It introduces a way of rendering additional fields on the UI via configuration. This is achieved by adding the following configuration to Kahuna service and Media API service configuration file i.e. 

- `$USER/.grid/kahuna.conf` or `/etc/grid/kahuna.conf` 
- `$USER/.grid/media-api.conf` or `/etc/grid/media-api.conf` 

where applicable (dev stage or prod stage respectively).

```hocon
# Field aliases configurations that are used to define which metadata field is visible, searchable and has an alias on the UI
field.aliases = [
  {
    # Sets the image metadata tag to be used when filtering FileMetadata field
    elasticsearchPath = "fileMetadata.xmp.bbc:Programme-Maker"
    # Sets an alias for use in search, shouldn't have whitespace or special characters
    alias = "BBCProgrammeMaker"
    # Sets a human readable label that will rendered on the UI
    label = "BBC Programme Maker"
    # Sets the field alias to be displayed as a search hint when you type `+`
    displaySearchHint = false
  }
  {
    elasticsearchPath = "fileMetadata.xmp.xmp:CreatorTool"
    alias = "creatorTool"
    label = "Creator Tool"
    displaySearchHint = false
  }
  {
    elasticsearchPath = "fileMetadata.exif.Artist"
    alias = "artist"
    label = "Artist"
    displaySearchHint = true
  }
  {
    elasticsearchPath = "fileMetadata.iptc.Caption Writer/Editor"
    alias = "captionWriter"
    label = "Caption Writer / Editor"
    displaySearchHint = true
  }
  {
    elasticsearchPath = "fileMetadata.xmp.Iptc4xmpExt:DigitalSourceType"
    alias = "digitalSourceType"
    label = "Digital Source Type"
    displaySearchHint = false
  }
]
```

The approach is described in the proposal: https://github.com/guardian/grid/issues/3167

## How can success be measured?

On the Image view page, the configured fields are visible on the “Show additional metadata” section when expanded.

## Screenshots

Screenshot at 2021-03-02 20-44-10

Screenshot at 2021-03-02 20-45-57

## Who should look at this?

@AWare , @sihil , @paperboyo and @akash1810 

## Tested?
- [x] locally by committer
- [ ] locally by Guardian reviewer
- [ ] on the Guardian's TEST environment

### CODE REVIEW

**Comment by ochiengolanga** [comment]
Sample image used:

sample-image

File: kahuna/app/lib/KahunaConfig.scala [hunk 1]
```diff
@@ -21,4 +23,22 @@ class KahunaConfig(resources: GridConfigResources) extends CommonConfig(resource
   val supportEmail: Option[String]= stringOpt("links.supportEmail").filterNot(_.isEmpty)
 
   val frameAncestors: Set[String] = getStringSet("security.frameAncestors")
+  val fileMetadataConfigs: List[FileMetadataConfig] = configObjectOpt("filemetadata.configurations") match {
+    case Some(config) => config.unwrapped()
+      .asInstanceOf[java.util.Map[String, java.util.HashMap[String, java.util.HashMap[String, Object]]]]
```
Comment (AWare) [inline kahuna/app/lib/KahunaConfig.scala:28]: Can we extract them without using `asInstanceOf`?

Comment (ochiengolanga) [reply kahuna/app/lib/KahunaConfig.scala:28]: No, from the structure of the HOCON config with dynamic paths. More than happy to get any pointers on how to achieve this without casting.

Comment (ochiengolanga) [reply kahuna/app/lib/KahunaConfig.scala:28]: The cast `.asInstanceOf[java.util.Map[String, java.util.HashMap[String, java.util.HashMap[String, Object]]]]` can be cleaner by casting to `ConfigObject` instead i.e. `asInstanceOf[ConfigObject]`

**Comment by sihil** [comment]
We'll do a code review shortly but wanted to put this here as a general comment on the structure of the configuration.

We think that the configuration can be more generic. Specifically:
- Rather than have discreet objects for each metadata type it can be only one
- `elasticsearchPath` replaces `tag` and contains the full path instead of the suffix (means that it can be used for any field in ES)
- the `alias` is used in searching and chips
- `label` is introduced to be the human readable name
- The whole config becomes an array instead of an object as the keys are redundant and it means that ordering in the UI can be derived from the order here

Taking your example:
- `xmp.bbc:Programme-Maker` as Visible and Searchable with an alias “BBCProgrammeMaker”
- `xmp.aux:Lens` as Visible with an alias “AuxLens”
- `xmp.CreatorTool` as visible with an alias “Creator Tool”
- `exif:Artist` as visible and searchable with an alias “Artist”

```hocon
# FileMetadata configurations that are used to define which metadata field is visible, searchable and has an alias on the UI
fieldAliases [
  {
    # Sets the image metadata tag to be used when filtering FileMetadata field
    elasticsearchPath = "fileMetadata.xmp.bbc:Programme-Maker"
    # Sets an alias for use in search, shouldn't have whitespace or special characters
    alias = "BBCProgrammeMaker"
    # Sets a human readable label that will rendered on the UI
    label = "BBC Programme Maker"
    # Sets the field alias to be displayed as a search hint when you type `+`
    displaySearchHint = false
  },
  {
    elasticsearchPath = "fileMetadata.xmp.aux:Lens"
    alias = "auxLens"
    label = "Aux Lens"
    displaySearchHint = false
  },
  {
    elasticsearchPath = "fileMetadata.xmp.CreatorTool"
    alias = "creatorTool"
    label = "Creator Tool"
    displaySearchHint = false
  },
  {
    elasticsearchPath = "fileMetadata.exif.Artist"
    alias = "artist"
    label = "Artist"
    displaySearchHint = true
  }
]
```

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js [hunk 2]
```diff
@@ -40,6 +40,28 @@ module.controller('grImageMetadataCtrl', [
     ctrl.metadata = ctrl.image.data.metadata;
     ctrl.identifiers = ctrl.image.data.identifiers;
 
+    mediaApi.fileMetadata(ctrl.image).then(resource => {
```
Comment (sihil) [inline kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js:43]: We probably don't want to read `fileMetadata` as a matter of course. This can be in excess of 1Mb in many cases and we'll only ever want a tiny fraction of the content. Either the field aliases should be automatically added to the main image endpoint or a new endpoint that only serves those values can be created.

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js [hunk 3]
```diff
@@ -40,6 +40,28 @@ module.controller('grImageMetadataCtrl', [
...
+      return resource.data;
+    }).then(fileMetadata => {
+      const groupedConfigsByDirectory = window._clientConfig.fileMetadataConfig.reduce((r, a) => {
```
Comment (sihil) [inline kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js:46]: Rather than read this here every time it is used it should be parsed once in a controller. e.g. https://github.com/guardian/grid/blob/3b03881f9aa9a6331bb34c99967712e842540998/kahuna/public/js/errors/global.js#L44

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js [hunk 4]
```diff
@@ -40,6 +40,28 @@ module.controller('grImageMetadataCtrl', [
...
+      let object = {};
+
+      Object.keys(groupedConfigsByDirectory).forEach(directory => {
+        groupedConfigsByDirectory[directory]
+          .filter(config => config.visible)
+          .filter(fileMetadataConfig => fileMetadata[directory][fileMetadataConfig.tag] !== undefined)
+          .map(fileMetadataConfig => {
+            object[fileMetadataConfig.alias] = fileMetadata[directory][fileMetadataConfig.tag];
+          });
+      });
```
Comment (sihil) [inline kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js:60]: A cleaner pattern for this (once flattened to remove the directories) uses `Object.fromEntries`. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries#object_transformations

File: common-lib/src/main/scala/com/gu/mediaservice/lib/config/FileMetadataConfig.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+object FileMetadataConfig {
+  import play.api.libs.json._
+  implicit val MetadataConfigurationWrites: Writes[FileMetadataConfig] = (
+    (__ \ "directory").write[String] ~
+      (__ \ "tag").write[String] ~
+      (__ \ "visible").write[Boolean] ~
+      (__ \ "searchable").write[Boolean] ~
+      (__ \ "alias").writeNullable[String]
+    )(unlift(FileMetadataConfig.unapply))
```
Comment (sihil) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/config/FileMetadataConfig.scala:20]: Think you can probably use a macro here?
```suggestion
  implicit val MetadataConfigurationWrites: Writes[FileMetadataConfig] = Json.writes[FileMetadataConfig]
```

File: common-lib/src/main/scala/com/gu/mediaservice/lib/config/FileMetadataConfig.scala [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+case class FileMetadataConfig(
+                               directory: String,
+                               tag: String,
+                               visible: Boolean = false,
+                               searchable: Boolean = false,
+                               alias: Option[String])
+
+object FileMetadataConfig {
+  import play.api.libs.json._
```
Comment (sihil) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/config/FileMetadataConfig.scala:13]: Idiomatically this import belongs at the top of the file.

**Review by sihil — COMMENTED** [review]
A few thoughts for you to look at. We're very happy to run through in a call if you have any questions or want to discuss further.

File: common-lib/src/main/scala/com/gu/mediaservice/lib/config/CommonConfig.scala [hunk 7]
```diff
@@ -94,6 +92,8 @@ abstract class CommonConfig(val configuration: Configuration) extends AwsClientB
   final def boolean(key: String): Boolean =
     configuration.getOptional[Boolean](key).getOrElse(false)
 
+  final def configObjectOpt(key : String) : Option[ConfigObject] = configuration.getOptional[ConfigObject](key)
```
[the remark this answers is not in the document]
Comment (ochiengolanga) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/config/CommonConfig.scala]: Thanks @sihil. Definitely a cleaner implementation now

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html [hunk 8]
```diff
@@ -272,14 +272,20 @@
                 ng-click="metadataExpanded = !metadataExpanded">
             <span ng-hide="metadataExpanded">▸ Show</span>
             <span ng-show="metadataExpanded">▾ Hide</span>
-            full metadata
+            image metadata
```
Comment (sihil) [inline kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html:275]: What's the motivation behind this change? Looking at the UI there is already metadata on display above this so show image metadata doesn't seem to make as much sense. I imagine that the objection of full is that there is more metadata not being displayed?

image

File: kahuna/app/views/main.scala.html [hunk 9]
```diff
@@ -43,7 +44,8 @@
           feedbackFormLink: "@Html(feedbackFormLink.getOrElse(""))",
           usageRightsHelpLink: "@Html(usageRightsHelpLink.getOrElse(""))",
           invalidSessionHelpLink: "@Html(invalidSessionHelpLink.getOrElse(""))",
-          supportEmail: "@Html(supportEmail.getOrElse(""))"
+          supportEmail: "@Html(supportEmail.getOrElse(""))",
+          fileMetadataConfig: @Html(fileMetadataConfigs)
```
Comment (sihil) [inline kahuna/app/views/main.scala.html:48]: I'm somewhat apprehensive of having a field here that isn't escaped but it looks like we have a number of other examples and it isn't coming from the user so it's relatively safe.

File: kahuna/public/js/services/api/media-api.js [hunk 10]
```diff
@@ -66,6 +66,10 @@ mediaApi.factory('mediaApi',
         return root.follow('image', {id: id}).get();
     }
 
+    function fileMetadata(image, {include} = {}) {
+      return image.follow('fileMetadata', { include: include }).get();
+    }
```
Comment (sihil) [inline kahuna/public/js/services/api/media-api.js:71]: do we need this anymore now we're not fetching `fileMetadata`?

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js (hunk 3 above)
Comment (ochiengolanga) [reply kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js]: Removed this bit that's now irrelevant. The visible fileMetadata fields are accessible within `ctrl.image` variable hence no need to fetch `fileMetadata` via the `mediaApi`

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js (hunk 4 above)
Comment (ochiengolanga) [reply kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js]: Removed this bit that's now irrelevant. The visible fileMetadata fields are accessible within `ctrl.image`

File: media-api/app/lib/ImageResponse.scala [hunk 11]
```diff
@@ -320,6 +324,28 @@ object ImageResponse {
...
+      }
+    }
+
+    val aliases: Seq[(String, JsValue)] = config.fieldAliasConfigs.flatMap { config =>
+      val parts = config.elasticsearchPath.split('.').toList.filter(_.nonEmpty)
+      val lookupResult = nestedLookup(JsDefined(source.source), parts)
+      lookupResult.toOption.map {
+        config.alias -> _
+      }
+    }
```
Comment (sihil) [inline media-api/app/lib/ImageResponse.scala:343]: Fairly minor but it would be pleasant to extract this into another method so that this business logic isn't cluttering this function. Perhaps something like `extractAliasFieldValues(config: MediaApiConfig, source: SourceWrapper[Image]): Seq[(String, JsValue)]` on the `ImageResponse` object.

In fact `addAliases` then becomes `addAliases(aliases: Seq[(String, JsValue)])` and the aliases are actually calculated at the call site around line 89.

**Review by sihil — COMMENTED** [review]
Morning @ochiengolanga. This PR is in good shape and very close to being ready. Comments are mainly tidying up this time around. With a little refactor of the `ImageResponse` code. Let me know if you have any questions.

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html (hunk 8 above)
Comment (paperboyo) [reply kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html:275]: I agree with @sihil. If `full` is no good, maybe we can go with `more`?

File: common-lib/src/main/scala/com/gu/mediaservice/model/FileMetadata.scala [hunk 12]
```diff
@@ -1,11 +1,9 @@
 package com.gu.mediaservice.model
 
 import com.gu.mediaservice.lib.logging.{LogMarker, MarkerMap}
-import net.logstash.logback.marker.Markers
-import play.api.libs.json._
 import play.api.libs.functional.syntax._
-
-import scala.collection.JavaConverters._
```
[the remark this answers is not in the document]
Comment (ochiengolanga) [reply common-lib/src/main/scala/com/gu/mediaservice/model/FileMetadata.scala:8]: Just optimizing imports here

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html (hunk 8 above)
Comment (ochiengolanga) [reply kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html:275]: This change was inspired by a suggestion on the [proposal](https://github.com/guardian/grid/issues/3167) to change from "Show full metadata" to "Show image metadata".

Comment (paperboyo) [reply kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html:275]: My suggestion makes no sense: `Hide more metadata` sounds idiotic. Akash, Alex and me think `additional` will work and is better than `full` (not really true).

Comment (ochiengolanga) [reply kahuna/public/js/components/gr-image-metadata/gr-image-metadata.html:275]: Changing the text to "Show additional metadata" and "Hide additional metadata". Thanks @paperboyo

File: common-lib/src/main/scala/com/gu/mediaservice/model/FileMetadata.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (sihil) [reply common-lib/src/main/scala/com/gu/mediaservice/model/FileMetadata.scala:8]: Understood but I'd avoid doing so in a file you have not otherwise touched.

File: media-api/test/lib/ImageResponseTest.scala [hunk 13]
```diff
@@ -1,11 +1,15 @@
...
+import com.gu.mediaservice.model._
 import com.gu.mediaservice.model.usage.{PendingUsageStatus, PrintUsage, Usage}
-import com.gu.mediaservice.model.{Bounds, Crop, CropSpec}
+import lib.elasticsearch.{ElasticSearchTestBase, SourceWrapper}
 import org.joda.time.DateTime.now
+import org.scalatest.Inside.inside
 import org.scalatest.{FunSpec, Matchers}
+import play.api.libs.json._
+
+class ImageResponseTest extends FunSpec with Matchers with ElasticSearchTestBase {
```
Comment (sihil) [inline media-api/test/lib/ImageResponseTest.scala:11]: Unless I've misunderstood this test no longer uses the ElasticSearchTestBase?

Comment (ochiengolanga) [reply media-api/test/lib/ImageResponseTest.scala]: Refactored `ImageResponseTest` and `ElasticSearchTestBase` to resolve this.

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js [hunk 14]
```diff
@@ -33,8 +33,20 @@ module.controller('grImageMetadataCtrl', [
...
     ctrl.showUsageRights = false;
     ctrl.usageRights = imageService(ctrl.image).usageRights;
+    ctrl.additionalMetadata = Object.fromEntries(
+      Object.entries(ctrl.image.data.aliases)
+        .map(([key, val]) => {
+          let match = ctrl.fieldAliases.find(_ => _.alias === key);
+          if (match) {
+            return [match.label, val];
+          } else {
+            return undefined;
```
Comment (sihil) [inline kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js:46]: Would it be better to fall back to `[key, val]` here? we're not expecting it to happen, but it would IMHO be better to show an alias instead of a label than drop the field completely.

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js [hunk 15]
```diff
@@ -33,8 +33,20 @@ module.controller('grImageMetadataCtrl', [
...
+          }
+        })
+        .filter(_ => _ !== undefined));
```
Comment (sihil) [inline kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js:49]: If you takethe suggestion above then this filter can also go.

File: common-lib/src/main/scala/com/gu/mediaservice/model/Image.scala [hunk 16]
```diff
@@ -127,4 +127,3 @@ object Image {
     )(unlift(Image.unapply))
 
 }
-
```
Comment (sihil) [inline common-lib/src/main/scala/com/gu/mediaservice/model/Image.scala:130]: Last bit of noise here.

**Review by sihil — APPROVED** [review]
This is a really nice piece of additional functionality. Thanks @ochiengolanga!

This is fundamentally ready to go. I left a couple of minor comments but I'd be OK merging it as it is.

Please do go ahead and squash your commits and then I'll look to merge shortly after that.

File: kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js (hunk 14 above)
Comment (ochiengolanga) [reply kahuna/public/js/components/gr-image-metadata/gr-image-metadata.js:46]: Definitely makes sense to fallback to alias, thanks!

**Comment by sihil** [comment]
Hi @ochiengolanga,
Have confirmed that this squash is good and will merge in the morning. Thanks again for this PR!

**Review by sihil — APPROVED** [review]
Tested at Guardian towers. Two minor points around the labels losing their casing and formatting of array values but we'll 🚢 this and raise an issue to resolve those.

image
